File: tests/Unit/PointwiseFunctions/Hydro/Test_MassFlux.cpp
// Distributed under the MIT License.
// See LICENSE.txt for details.

#include "Framework/TestingFramework.hpp"

#include <cstddef>
#include <string>

#include "DataStructures/DataBox/DataBox.hpp"
#include "DataStructures/DataVector.hpp"
#include "DataStructures/Tensor/Tensor.hpp"
#include "Framework/CheckWithRandomValues.hpp"
#include "Framework/SetupLocalPythonEnvironment.hpp"
#include "Helpers/DataStructures/DataBox/TestHelpers.hpp"
#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
#include "PointwiseFunctions/Hydro/MassFlux.hpp"
#include "PointwiseFunctions/Hydro/Tags.hpp"
#include "Utilities/TMPL.hpp"

namespace hydro {
namespace {
template <size_t Dim, typename Frame, typename DataType>
void test_mass_flux(const DataType& used_for_size) {
  pypp::check_with_random_values<1>(
      static_cast<tnsr::I<DataType, Dim, Frame> (*)(
          const Scalar<DataType>&, const tnsr::I<DataType, Dim, Frame>&,
          const Scalar<DataType>&, const Scalar<DataType>&,
          const tnsr::I<DataType, Dim, Frame>&,
          const Scalar<DataType>&) noexcept>(&mass_flux<DataType, Dim, Frame>),
      "TestFunctions", "mass_flux", {{{-10.0, 10.0}}}, used_for_size);
}
}  // namespace

SPECTRE_TEST_CASE("Unit.PointwiseFunctions.Hydro.MassFlux",
                  "[Unit][Hydro]") {
  pypp::SetupLocalPythonEnvironment local_python_env(
      "PointwiseFunctions/Hydro/");
  const DataVector dv(5);
  test_mass_flux<1, Frame::Inertial>(dv);
  test_mass_flux<1, Frame::Grid>(dv);
  test_mass_flux<2, Frame::Inertial>(dv);
  test_mass_flux<2, Frame::Grid>(dv);
  test_mass_flux<3, Frame::Inertial>(dv);
  test_mass_flux<3, Frame::Grid>(dv);

  test_mass_flux<1, Frame::Inertial>(0.0);
  test_mass_flux<1, Frame::Grid>(0.0);
  test_mass_flux<2, Frame::Inertial>(0.0);
  test_mass_flux<2, Frame::Grid>(0.0);
  test_mass_flux<3, Frame::Inertial>(0.0);
  test_mass_flux<3, Frame::Grid>(0.0);

  // Check compute item works correctly in DataBox
  TestHelpers::db::test_compute_tag<
      Tags::MassFluxCompute<DataVector, 2, Frame::Inertial>>("MassFlux");
  Scalar<DataVector> rho{{{DataVector{5, 1.0}}}};
  tnsr::I<DataVector, 3> velocity{
      {{DataVector{5, 0.25}, DataVector{5, 0.1}, DataVector{5, 0.35}}}};
  Scalar<DataVector> lorentz{{{DataVector{5, 0.2}}}};
  Scalar<DataVector> lapse{{{DataVector{5, 0.3}}}};
  tnsr::I<DataVector, 3> shift{
      {{DataVector{5, 0.1}, DataVector{5, 0.2}, DataVector{5, 0.3}}}};
  Scalar<DataVector> sqrt_det_g{{{DataVector{5, 0.25}}}};
  const auto box = db::create<
      db::AddSimpleTags<Tags::RestMassDensity<DataVector>,
                        Tags::SpatialVelocity<DataVector, 3, Frame::Inertial>,
                        Tags::LorentzFactor<DataVector>,
                        ::gr::Tags::Lapse<DataVector>,
                        ::gr::Tags::Shift<3, Frame::Inertial, DataVector>,
                        ::gr::Tags::SqrtDetSpatialMetric<DataVector>>,
      db::AddComputeTags<
          Tags::MassFluxCompute<DataVector, 3, Frame::Inertial>>>(
      rho, velocity, lorentz, lapse, shift, sqrt_det_g);
  CHECK(db::get<Tags::MassFlux<DataVector, 3, Frame::Inertial>>(box) ==
        mass_flux(rho, velocity, lorentz, lapse, shift, sqrt_det_g));
}
}  // namespace hydro

File: tests/Unit/PointwiseFunctions/Hydro/Python/Test_MassFlux.py
# Distributed under the MIT License.
# See LICENSE.txt for details.

import os
import sys
import unittest

import numpy as np
import numpy.testing as npt

import spectre.DataStructures.Tensor.Frame as fr
import spectre.PointwiseFunctions.Hydro as hydro
from spectre import Informer
from spectre.DataStructures import DataVector
from spectre.DataStructures.Tensor import Scalar, tnsr

sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))
from TestFunctions import *


class TestMassFlux(unittest.TestCase):
    def test_mass_flux(self):
        rest_mass_density = Scalar[DataVector](num_points=5, fill=2.1)
        spatial_velocity = tnsr.I[DataVector, 3, fr.Grid](
            num_points=5, fill=-1.0
        )

        lorentz_factor = Scalar[DataVector](num_points=5, fill=2.1)
        shift = tnsr.I[DataVector, 3, fr.Grid](num_points=5, fill=1.5)
        lapse = Scalar[DataVector](num_points=5, fill=1.1)
        sqrt_det_spatial_metric = Scalar[DataVector](num_points=5, fill=1.8)

        bindings = hydro.mass_flux(
            rest_mass_density,
            spatial_velocity,
            lorentz_factor,
            lapse,
            shift,
            sqrt_det_spatial_metric,
        )

        alternative = mass_flux(
            np.array(rest_mass_density)[0],
            np.array(spatial_velocity),
            np.array(lorentz_factor)[0],
            np.array(shift),
            np.array(lapse)[0],
            np.array(sqrt_det_spatial_metric)[0],
        )

        assert type(bindings) == tnsr.I[DataVector, 3, fr.Grid]
        np.testing.assert_allclose(bindings, alternative)


if __name__ == "__main__":
    unittest.main(verbosity=2)
